src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/filesystem/BitbucketSCMFileSystem.java

```diff
@@ -0,0 +1,143 @@
+import jenkins.scm.api.*;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nullable;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static org.eclipse.jgit.lib.Constants.*;
+
+public class BitbucketSCMFileSystem extends SCMFileSystem {
+
+    private final BitbucketFilePathClient client;
+    private final String ref;
+
+    protected BitbucketSCMFileSystem(BitbucketFilePathClient client, @Nullable SCMRevision scmRevision,
+                                     @Nullable String ref) {
+        super(scmRevision);
+        this.client = client;
+        this.ref = ref;
+    }
+
+    @Override
+    public SCMFile getRoot() {
+        return new BitbucketSCMFile(client, ref);
+    }
+
+    // We do not provide this information in the REST response, so this is undefined.
+    @Override
+    public long lastModified() throws IOException, InterruptedException {
+        return 0;
+    }
+
+    @Extension
+    public static class BuilderImpl extends SCMFileSystem.Builder {
+
+        @Inject
+        BitbucketClientFactoryProvider clientFactoryProvider;
+        @Inject
+        JenkinsToBitbucketCredentials jenkinsToBitbucketCredentials;
+        @Inject
+        BitbucketPluginConfiguration pluginConfiguration;
+
+        @Override
+        public SCMFileSystem build(Item item, SCM scm,
+                                   @CheckForNull SCMRevision scmRevision) throws IOException, InterruptedException {
+            if (!(scm instanceof BitbucketSCM)) {
+                return null;
+            }
+
+            BitbucketSCM bitbucketSCM = (BitbucketSCM) scm;
+            Optional<BitbucketServerConfiguration> maybeServerConfiguration =
+                    pluginConfiguration.getServerById(bitbucketSCM.getServerId());
+            if (!maybeServerConfiguration.isPresent() || maybeServerConfiguration.get().validate().kind == Kind.ERROR) {
+                return null;
+            }
+
+            BitbucketSCMRepository repository = bitbucketSCM.getBitbucketSCMRepository();
+
+            BitbucketFilePathClient filePathClient =
+                    clientFactoryProvider.getClient(maybeServerConfiguration.get().getBaseUrl(),
+                            jenkinsToBitbucketCredentials.toBitbucketCredentials(repository.getCredentialsId()))
+                            .getProjectClient(repository.getProjectKey())
+                            .getRepositoryClient(repository.getRepositorySlug())
+                            .getFilePathClient();
+
+            return new BitbucketSCMFileSystem(filePathClient, null, bitbucketSCM.getBranches().get(0).toString());
+        }
+
+        @Override
+        public SCMFileSystem build(SCMSource source, SCMHead head,
+                                   @CheckForNull SCMRevision scmRevision) throws IOException, InterruptedException {
+            if (!(source instanceof BitbucketSCMSource)) {
+                return null;
```

Comment:
I spent a bit of time thinking about this, and although I agree it's weird, our hands are tied by how the SPI works and who consumes it. You can see the flow in `CpsScmFlowDefinition.` for pipelines and `org.jenkinsci.plugins.workflow.multibranch.SCMBinder` for multibranch, and notice in neither case is there elegant exception handling. If we return a `null` for this builder, the job is considered lightweight unsupported and proceeds with a full checkout. If we throw an exception, the job fails with a stack trace in the logging. Although there are certainly cases where it might be advantageous for the build to fail _before_ performing a full checkout (e.g. a server configuration with no base URL, where the build will fail anyway), killing the build with a runtime exception feels a very blunt tool to use.
There is some prior art to consider- both the `GitSCMFileSystem` (in the git plugin) and `BitbucketBranchSourceFileSystem` (in branch source) also return nulls here. In the case of the pipeline job we can potentially add some logging to say why the job failed, and I _think_ we can do this for the branch source as well- I think that'd be a nice addition, but I don't think we should throw exceptions.

Reply:
Thanks for the information! I feel like I'm still lacking a bit of context though - who/what are the consumers of the `BitbucketSCMFileSystem#builder`? Is this going to be used by jenkins?
If the convention is to return null for builders in jenkins then i can see why you've included it here

Reply:
Okay I was wrong, we can't use job logging- there's no way to get the task listener. I guess users will just have to guess why their lightweight checkout failed.
/tableflip

Reply:
:(